### ISSUE TITLE
Loader would lost bytes and be disability if GC untimely

### ISSUE THREAD

**Issue by xiaoChi888** [airsdk/Adobe-Runtime-Support#2731]
## Problem Description

AIR SDK version: 50.2.2.6
OS: Windows 10 (not sure if any other OS effected)

## Steps to Reproduce
Having tested, there's a simple version ActionScript code can reproduce the bug:

	package {
		import flash.display.MovieClip;
		import flash.display.Loader;
		import flash.net.URLRequest;
		import flash.events.Event;
		import flash.events.ProgressEvent;
		import flash.display.LoaderInfo;
		import flash.events.IOErrorEvent;
		import flash.system.System;
		import flash.utils.Timer;
		import flash.events.TimerEvent;
	
		public class MainTimeLine extends MovieClip {
			public const gcTimer: Timer = new Timer(2000); //With different rate, the issue may not appear
			public const loader: Loader = new Loader();
			private var bool: Boolean = false;
	
			public function MainTimeLine() {
				addChild(loader);
				first();
			}
			private function first() {
				gcTimer.addEventListener(TimerEvent.TIMER, garbageCollect);
				loader.contentLoaderInfo.addEventListener(ProgressEvent.PROGRESS, outputBytesLoaded);
				loader.contentLoaderInfo.addEventListener(Event.COMPLETE, second);
				loader.contentLoaderInfo.addEventListener(IOErrorEvent.IO_ERROR, loadFailed);
				gcTimer.start();
				loader.load(new URLRequest(getResourceURLWith(1000)));
			}
			private function second(e: Event) {
				loader.contentLoaderInfo.removeEventListener(Event.COMPLETE, second);
				loader.unload();
				loader.load(new URLRequest(getResourceURLWith(1001))); //It returns a URL to an empty resource
			}
			private function third() {
				loader.unload();
				loader.load(new URLRequest(getResourceURLWith(1002)));
			}
			private function outputBytesLoaded(e: ProgressEvent) {
				trace("Loaded " + loader.loaderInfo.bytesLoaded + "B (" + loader.loaderInfo.url + ")");
			}
			private function loadFailed(e: IOErrorEvent) {
				trace("Load failed");
			}
			private function getResourceURLWith(id: uint): String {
				return "https://seer.61.com/resource/fightResource/pet/swf/" + String(id) + ".swf";
			}
			private function garbageCollect(e: TimerEvent) {
				System.gc();
				trace("Loaded " + loader.contentLoaderInfo.bytesLoaded + "B");
				if (loader.contentLoaderInfo.bytesLoaded == 0 && !bool) {
					third();
					bool = true;
				} else {
					gcTimer.reset();
					gcTimer.start();
				}
			}
		}
	
	}

But actually, i'm not sure if it's the same issue cause my program bug… `loaderInfo.bytesLoaded` returns 168 when loading a URL to an empty resource. After GC the 168B (the `loaderInfo.bytesLoaded` returns 0), the `Loader` can't load other URL and display the resource normally.

## Known Workarounds

Create a new Loader to replace the old one before load…

**Comment by ajwfrost** [comment]
Hi

I think the test case is a little confused, because the "third()" call is made before the first loader has got very far, so then you have two 'load' calls happening at once. If we take that out, then we see:
1) the first load is kicked off, and completes (progress event has 1500B/1500B, but then the final total bytes loaded are 41381)
2) the second load is kicked off and says it failed - but the loader has 168 bytes (which is the size of that file) - and we got a progress event saying 1500B / 1500B had loaded.

So it may more be that the re-use of the loader is causing a problem with the progress event and the expected total data size. We can see whether it's possible to sort this out by resetting everything if a new 'load' is called on a loader that already has an ongoing or completed load...

thanks

**Comment by xiaoChi888** [comment]
Yeah, it's a little confused… i write the sample code when i'm sleepy… i just want to reproduce the logic of my program that time, because it really faced the bug. But the sample may not reproduce it…
 
In my program, it is odd. Before the program do load, a variable will be set to `null`. It should gets the animation `MovieClip` with `new(domain.getDefinition("pet") as Class)()`. After GC when the `Loader` load with a URL to an empty resource, the `loaderInfo.bytesLoaded` change to 0 from 168. Then use the `Loader` to load with a new URL, everything seems going on but the variable didn't get the animation, what's more strange is that the variable still reserve a reference to the old animation…
 
Not sure if you have interested in the concrete program (it's a bit complicated after all). Here is the source file ([source file version with Loader bug.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/12050896/source.file.version.with.Loader.bug.zip)), The relevant logic is mainly in `FightPanel.as`, the function `loadPets()` do the behavior i talked about in the last paragraph. Click the right-bottom button "精灵" can open the load set panel, and the first `TextField` is the argument about URL (like in that simple code, 1000, 1001, 1002 can constitute test). The 4 buttons can call `gotoAndStop()` of the animation (maybe it can means the reference didn't disconnect?).
 
Thank you.

**Comment by xiaoChi888** [comment]
Well, for more test, it seems the `Loader` can load resource after GC. But there's still some issues in it. One is that after GC when `Loader` load an empty resource (the 168 bytes), the `loaderInfo.url` returns a wrong URL (normally it returns the resource's URL), but after the untimely GC, it returns the URL where the request be made (app:/main.swf). Another one is the `trace()` can output the class by `ApplicationDomian.getDefinition()`, but it didn't be successfully added to the display-object list (maybe some security sandbox error?).
 
Thank you.
